std/haxe/iterators/MapKeyValueIterator.hx

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (C)2005-2018 Haxe Foundation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+package haxe.iterators;
+
+import haxe.ds.IntMap;
+
+/**
+	This Key/Value iterator can be used to iterate across maps.
+**/
+@:ifFeature("anon_read.keyValueIterator", "dynamic_read.keyValueIterator")
+class MapKeyValueIterator<K, V> {
+	var map:haxe.Constraints.IMap<K, V>;
+	var keys:Iterator<K>;
+
+ 	public inline function new(map:haxe.Constraints.IMap<K, V>) {
+		this.map = map;
+		this.keys = map.keys();
+	}
+
+	/**
+		See `Iterator.hasNext`
+	**/
+ 	public inline function hasNext():Bool {
+		return keys.hasNext();
+	}
+
+	/**
+		See `Iterator.next`
+	**/
+ 	public inline function next():{key:K,value:V} {
+		var key = keys.next();
+		return {value: map.get(key), key:key};
```

Comment:
I believe, key-value iterator should be implemented per target.
With the proposed implementation you get invalid values for the keys, which were removed in previous steps.
```haxe
var map ['first' => 1, 'second' => 2];
for(key => value in map) {
  trace(key, value);
  map.remove('second');
}
```
This example will emit
```
first,1
second,null
```
While it should be either
```
first,1
second,2
```
or
```
first,1
```

Reply:
You can't modify while iterate on arrays either.

Reply:
You can )
https://try.haxe.org/#eD62c

Reply:
It only works like that because the loop doesn't cache the array `.length` access. That's a luxury we have on Array because we can assume it to be an O(1) operation. We can't do something like that on maps though.

Reply:
We could try to utilize native key/value iteration loops on each platform.

Reply:
I'm not sure that is worth the effort.

Let's go for the current approach for now.

Reply:
But changing it then will be a breaking change...

Reply:
Modify-while-iterate is undefined behavior.

Also, not all targets could support this natively, so there would be an inconsistency either way.